tutorials/Tutorial MIS & QAOA.ipynb

```diff
@@ -0,0 +1,394 @@
+    "    seq.declare_channel('ch0','rydberg_global')\n",
+    "    middle = int(len(param)/2)\n",
+    "    param = np.array(param)*1 #wrapper \n",
+    "    t = param[:middle] #associated to H_c\n",
+    "    tau = param[middle:] #associated to H_0\n",
+    "    p = len(t)\n",
+    "    for i in range(p):\n",
+    "        pulse_1 = Pulse.ConstantPulse(tau[i], 1., 0, 0) # H_M \n",
+    "        pulse_2 = Pulse.ConstantPulse(t[i], 1., 1., 0) # H_M + H_c \n",
+    "        seq.add(pulse_1, 'ch0')\n",
+    "        seq.add(pulse_2, 'ch0')\n",
+    "    seq.measure('ground-rydberg')\n",
+    "    simul = Simulation(seq, skip=100)\n",
+    "    results = simul.run()\n",
+    "    count_dict = results.sample_final_state(N_samples=1000) #sample from the state vector \n",
+    "    return count_dict "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "example_dict = quantum_loop([300,900], r = reg)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can then plot the distribution of the samples, to see the most frequent bitstrings sampled."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def plot_distribution(C):\n",
+    "    C = dict(sorted(C.items(), key=lambda item: item[1], reverse=True))\n",
+    "    color_dict = {key: 'g' for key in C}\n",
+    "    indexes = ['01011', '00111']  # MIS indexes\n",
+    "    for i in indexes: \n",
+    "        color_dict[i] = 'red'\n",
+    "    plt.figure(figsize=(12,6))\n",
+    "    plt.xlabel(\"bitstrings\")\n",
+    "    plt.ylabel(\"counts\")\n",
+    "    plt.bar(C.keys(), C.values(), width=0.5, color = color_dict.values())\n",
+    "    plt.xticks(rotation='vertical')\n",
+    "    plt.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plot_distribution(example_dict)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The bitstrings '01011' and '00111' (in red) correspond to the two MIS of the graph. The goal of QAOA is to choregraph interferences between the basis states, in order to maximize the frequency of the MIS states. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Optimization "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We estimate the cost of a sampled state vector by making an average over the samples. Determining the cost of a given bitstring takes polynomial time. The average estimate is then used in the classical loop to optimize the variational parameters $\\tau$ and $t$."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_cost_colouring(z,G,penalty=10):\n",
```

Comment:
Question: is the value of the `penalty` parameter related to the interaction strength that defined the register (which was set to "1" ?)

Reply:
While experimentally it would indeed correspond to the interaction strength, the idea here is to tune the penalty value as you want to strongly/weakly differentiate the computational basis states.